### PR TITLE
Feat add cellRenderer

### PR DESCRIPTION
**Pull request by Wroud**
#2937

### CODE REVIEW

File: src/Cell.tsx [hunk 1]
```diff
@@ -25,31 +25,36 @@ const cellDraggedOver = css`
...
+    rowIdx,
+    className,
+    onClick,
+    onDoubleClick,
+    onContextMenu,
+    onRowChange,
+    selectCell,
+    ...props
+  }: CellRendererProps<R, SR>,
+  refComponent: React.Ref<HTMLDivElement>
```
Comment (amanmahajan7) [inline src/Cell.tsx:45]:
```suggestion
  ref: React.Ref<HTMLDivElement>
```

File: src/Cell.tsx [hunk 2]
```diff
@@ -95,6 +100,7 @@ function Cell<R, SR>({
       aria-colspan={colSpan}
       aria-selected={isCellSelected}
       aria-readonly={!isEditable || undefined}
+      ref={refComponent}
```
Comment (amanmahajan7) [inline src/Cell.tsx:103]:
```suggestion
      ref={ref}
```

File: src/Row.tsx [hunk 3]
```diff
@@ -33,6 +34,9 @@ function Row<R, SR>(
   }: RenderRowProps<R, SR>,
   ref: React.Ref<HTMLDivElement>
 ) {
+  const defaultComponents = useDefaultRenderers<R, SR>();
+  const cellRenderer = defaultComponents?.renderCell ?? defaultRenderCell;
```
Comment (amanmahajan7) [inline src/Row.tsx:38]:
```suggestion
  const renderCell = useDefaultRenderers<R, SR>().renderCell!;
```
This should not be undefined here

File: src/Row.tsx [hunk 4]
```diff
@@ -68,21 +72,20 @@ function Row<R, SR>(
...
-          isCopied={copiedCellIdx === idx}
-          isDraggedOver={draggedOverCellIdx === idx}
-          isCellSelected={isCellSelected}
-          onClick={onCellClick}
-          onDoubleClick={onCellDoubleClick}
-          onContextMenu={onCellContextMenu}
-          onRowChange={handleRowChange}
-          selectCell={selectCell}
-        />
+        cellRenderer(column.key, {
```
Comment (amanmahajan7) [inline src/Row.tsx:75]:
```suggestion
        renderCell(column.key, {
```

**Review by amanmahajan7 — COMMENTED** [review]
We also need tests for the `renderCell` prop

File: test/renderers.test.tsx [hunk 5]
```diff
@@ -174,3 +199,24 @@ test('sortPriority defined using both providers and renderers', async () => {
...
+
+test('renderCell defined using provider', () => {
+  setupProvider({ columns, rows: [{ id: 1, col1: 'col 1 value', col2: 'col 2 value' }] });
+
+  const [, cell1, cell2] = getCells();
+  expect(cell1).toHaveTextContent('col 1 value');
+  expect(cell2).toHaveTextContent('col 2 value');
+});
+
+test('renderCell defined using both providers ans renderers', () => {
```
Comment (amanmahajan7) [inline test/renderers.test.tsx:211]:
```suggestion
test('renderCell defined using both providers and renderers', () => {
```

File: README.md [hunk 6]
```diff
@@ -213,6 +213,7 @@ interface Renderers<TRow, TSummaryRow> {
   renderCheckbox?: Maybe<(props: RenderCheckboxProps) => ReactNode>;
   renderRow?: Maybe<(key: Key, props: RenderRowProps<TRow, TSummaryRow>) => ReactNode>;
   renderSortStatus?: Maybe<(props: RenderSortStatusProps) => ReactNode>;
+  renderCell?: Maybe<(key: Key, props: CellRendererProps<TRow, TSummaryRow>) => ReactNode>;
```
Comment (amanmahajan7) [inline README.md:216]:
```suggestion
  renderCell?: Maybe<(key: Key, props: CellRenderProps<TRow, TSummaryRow>) => ReactNode>;
```

Maybe we can rename now

File: src/index.ts [hunk 7]
```diff
@@ -2,6 +2,7 @@ export { default, type DataGridProps, type DataGridHandle } from './DataGrid';
 export { default as TreeDataGrid, type TreeDataGridProps } from './TreeDataGrid';
 export { DataGridDefaultRenderersProvider } from './DataGridDefaultRenderersProvider';
 export { default as Row } from './Row';
+export { default as Cell } from './Cell';
```
Comment (amanmahajan7) [inline src/index.ts:5]: Let's also export `RenderCellProps`

File: README.md (hunk 6 above)
Comment (Wroud) [reply README.md:216]: It conflicts with existing `RenderCellProps` used for `renderCell` property of `Column`

File: src/index.ts (hunk 7 above)
Comment (Wroud) [reply src/index.ts:5]: `RenderCellProps`
`CellRendererProps`
both re-exported from `types.ts`

**Comment by Liooo** [comment]
@amanmahajan7 @nstepien @Wroud 

What's the status on this PR? 
I'd love this feature, I'm happy to do the rest of work if there's anything left.

**Comment by amanmahajan7** [comment]
Completed in https://github.com/adazzle/react-data-grid/pull/3621

**Pull request closed without merging**